README
---------------------------------
Java Swing AlarmClock Application
=================================

Author - Copyright (C) 2012-2020 Mark Grant

Contents
========

1 ... Project Description

2 ... This Repository

3 ... Project Structure

4 ... Installation

		A) Just using the built jar file.
		B) For systems supporting AutoTools.
		C) Of Distro-Native Package
		D) Windows msi installer.

5 ... Utility Scripts


-------------------------

1 ... Project Description
=========================

This is a GUI AlarmClock application implementing two independent alarms. Each
alarm can be set to 'Alarm At', eg 13:15, or 'Alarm In', eg 1 hour 10 mins.
Each alarm also incorporates a 'Snooze' button that will pause a ringing alarm
for 5 minutes.


---------------------

2 ... This Repository
=====================

AlarmClock is written in Java using NetBeans as an IDE. The project must be
maintained via NetBeans to preserve it's support files, (*.form, build.xml,
etc) which are tracked in git. The repository is also configured for Autotools.
The AutoTools files exist in parallel to the NetBeans files and do not conflict.


-----------------------
3 ... Project Structure
=======================

The NetBeans part of the project is expected to be used for development only.
Once development is complete AutoTools takes over to provide basic build and
packaging which downstream can be turned into debian, rpm etc packages.
AutoTools will build an end-user application consisting of:-

	The main application jar file.
	A bash start-up script.
	A manual page for the start-up script.
	A desktop file for a GUI start-up of the application. (It invokes the
	bash script).


------------------

4 ... Installation
==================

As a Java application it can run on many operating systems.

A.1) Using the NetBeans build output
------------------------------------
The easiest way to use a jar file is to compile and build within NetBeans. This
produces the application in the dist subdirectory. The application can then be
run using:-

	java -jar "dist/AlarmClock.jar"
		with the path modified as necessary.
N.B. If you wish to move the application to another location you *must* move the
relative subdirectory 'lib' with it. i.e. the jar and the lib directory must
maintain their relative relationship.

A.2) Using a tarball of the NetBeans build output
-------------------------------------------------
To simplify the distribution of the jar file and lib directory as mentioned
above, a tarball can be created, distributed, extracted and used.

Create the archive:-

	./credisttar.sh
		with the ./ path modified as necessary.

Distribute the tarball created in project_root/distribution.

Extract the tarball in the desired location:-

	tar xvpzf ./AlarmClockvXXXXXXXXXX.tar.gz
		replacing XXXXXXXXXX with the embedded version string in the
		tarball name.

And then, as above, the application can be run using:-

	java -jar Alarmclock.jar

A.3) Using just the jar from the NetBeans build output
------------------------------------------------------
In order to just use the jar file, the supplemental jar libraries in the
NetBeans dist/lib directory (e.g. getopt) must already be installed on your
system.
Move the jar file to the desired location and run the command:-

	java -cp /usr/share/java/*:./AlarmClock.jar alarmclock.AlarmClock
		with the ./ path modified as necessary.

A.4) Using the jar built by AutoTools
-------------------------------------
In order to use the jar file built by AutoTools, the supplemental jar libraries
(e.g. getopt) must already be installed on your system.
Assuming you are in a build directory one level below project root, build the
project by running:-

	../bootstrap.sh -cba ..
Move the jar to the desired location and run the command:-

	java -cp /usr/share/java/*:./alarmclock-java-x.y.z.jar \
	alarmclock.AlarmClock
	where x.y.z will be the package version number. Modify the ./ path as
	necessary.

B)  AutoTools (configure and make) Installation
-----------------------------------------------
1 ... Download either the source or distribution tarball (the .tar.gz file)
	from:-

https://github.com/m-grant-prg/alarmclock-java/releases

2 ... Extract the tarball preserving the directory structure.

3 ... cd to the directory created.

4 ... If you downloaded the source tarball type 'autoreconf -if'

5(a) ... For Debian and closely related distros where support jar files are
located in /usr/share/java:-

		Type './configure --enable-atonly=yes'
		Now skip to step (6).

5(b) ... For distros which do not store support jar files in /usr/share/java:-

		Find the '/path/to/jars'
		Type './configure STDJARLOC=/path/to/jars --enable-atonly=yes'

6 ... As root or sudo, type 'make install clean'

(Quote marks are for textual clarity only).


To uninstall the package:

1 ... cd to the directory created in the above install process.

2 ... As root or sudo, type 'make uninstall clean'

C) Installation of Distro-Native Package
----------------------------------------
Installation packages native to different distributions are available, please
refer to the relevant installation section on the wiki at:-

https://github.com/m-grant-prg/alarmclock-java/wiki

D) Windows installation
-----------------------
Another associated GitHub repository AlarmClockWindowsDist is available. This
repository provides the mechanism to produce a Windows msi installer package.


---------------------
5 ... Utility Scripts
=====================
In the project root directory there are 2 helper scripts; bootstrap.sh and
credisttar.sh.

bootstrap.sh
------------
This misleadingly named script bootstraps the project build and provides other
useful features. The main options are probably b, c, C, D and T.

In AutoTools it is usually advisable to perform parallel builds. This means you
build somewhere other than the project root. This is because building creates
files and they would confuse the project root downwards. I always create a build
directory straight off the project root, cd to there and do all build and git
work from there, (.gitignore is already set to ignore such a directory).

Assuming you adopt the preceding paragraph then a typical invocation of the
script would be:-

../bootstrap.sh --config --build ..

The last '..' points the way to project root.

For the full list of arguments to bootstrap.sh, please refer to the options
section of the acmbuild wiki, the options are identical:-

https://github.com/m-grant-prg/acmbuild/wiki

credisttar.sh
-------------
This script produces a tarball from the NetBeans build output. The tarball can
then be used for distribution. Both project jar (eg AlarmClock.jar) and
libraries (eg getopt) are archived.

The script always produces the tarball in the project_root/distribution
directory and can be invoked from anywhere, eg:-

../credisttar.sh



NEWS
Release 1.3.7 2022-03-08
Add new config file to hold scaling_factor parameter. If your application
window looks tiny then your system is probably HiDPI and you should change the
scaling_factor. Starting with 2.0 is a good idea.

Release 1.3.6 2021-12-01
Correct license in Help:About dialog box

Release 1.3.5 2021-11-30
No user visible changes.
Technical change information can be found in the ChangeLog.

Release 1.3.4 2021-11-29
No user visible changes.
Technical change information can be found in the ChangeLog.

Release 1.3.3 2021-07-22
Technical changes only. Details can be found in the ChangeLog.

Release 1.3.2 2020-02-02
at: Bump minimum required java version to 11.0
Other technical only changes. Details can be found in the ChangeLog.

Release 1.3.1 2020-01-26
Update visible copyright dates
Other technical only changes. Details can be found in the ChangeLog.

Release 1.2.0 2020-01-24
Release 1.2.x series now pinned at stretch for java 8
java: Increase update timer frequency
at: netbeans: Correct jar object usage
Other technical only changes. Details can be found in the ChangeLog.

Release 1.1.5 2019-10-14
Technical changes only. Details can be found in the ChangeLog.

Release 1.1.4 2019-06-27
at: Update bootstrap.sh to v1.4.5
at: README: Improve AT installation instructions
at: configure.ac: Improve in-file comments
at: Fix make distcheck using enable-distcheckfake
at: Add classpath to javadoc to find getopt jar
travis: Setup Travis CI build testing
README: Point bootstrap.sh part at acmbuild wiki
git: Update project root gitignore to v1.0.2
Setup trap handling ASAP
Improve getopt CL processing

Release 1.1.3 2019-03-30
Fix javadoc creation errors with later versions.

Release 1.1.2 2019-03-30
Build javadoc to aid debugging.
Correct README formatting.
Update bootstrap.sh to v1.4.1

Release 1.1.1 2019-03-10
Correct Name and Generic Name usage in the desktop file.
Use standard categories from Freedesktop Desktop in desktop file.
Use wildcard in ClassPath for build and script
Improve comments in configure.ac CLA section
configure: Support STDJARLOC environment variable
Update man page exit status section
configure.ac: Adopt new standard layout
README: Add --enable-atonly to configure command
README: Update section on bootstrap.sh
README: Make package install references generic
Install standard GNU files eg AUTHOR etc
README: Add Copyright statement
Correct package versioning
Add bash exit code explanation
Pass bash path to scripts and Makefile.am's
Remove unnecessary shebangs from Makefile.am's
Update Copyright statement format
Use local and readonly variables where appropriate
Convert remaining echo statements to printf
Improve trap handling
Change [ x -gt y ] to (( .. ))
Add missing error handling after getopt
Remove script_exit_code variable
Update in-file ChangeLog
Standardise output() with (( )) and printf
Update to java-vm m4 macros v1.0.5
Add --enable-testinghacks to configure CLA
Add -a or --at-only build option
Update bootstrap.sh
alarmclock: Refactor CL processing into function
alarmclock: Refactor help text into function
alarmclock: Make loops and ifs more C-like
alarmclock: Do not hardcode outputprefix
Correct name capitalisation in bash script

Release 1.0.20 2018-10-27
srctarball to use -src- in filename

Release 1.0.19 2018-10-20
Add markdown symlink to README.
Beautify README for GitHub markdown presentation.
Modify README to refer to new wiki.

Release 1.0.18 2018-08-28
Maintenance release of project infrastructure changes only. No changes to
functionality.

Release 1.0.17 2018-07-25
Initial AutoTools release.



bootstrap.sh
#! /bin/bash
#########################################################################
#									#
#	bootstrap.sh is automatically generated,			#
#		please do not modify!					#
#									#
#########################################################################

#########################################################################
#									#
# Script ID: bootstrap.sh						#
# Author: Copyright (C) 2014-2019, 2021  Mark Grant			#
#									#
# Released under the GPLv3 only.					#
# SPDX-License-Identifier: GPL-3.0-only					#
#									#
# Purpose:								#
# To simplify the AutoTools distribution build.				#
#									#
# Syntax:	bootstrap.sh	[ -a || --at-only ] ||			#
#				[ -b || --build ] ||			#
#				[ -c || --config ] ||			#
#				[ -C || --distcheck ] ||		#
#				[ -d || --debug ] ||			#
#				[ -D || --dist ] ||			#
#				[ -g || --gnulib ] ||			#
#				[ -h || --help ] ||			#
#				[ -H || --header-check ] ||		#
#				[ -K || --check ] ||			#
#				[ -m || --menu-config ] ||		#
#				[ -p || --parallel-jobs ] ||		#
#				[ -s || --sparse ] ||			#
#				[ -t || --testing-hacks ] ||		#
#				[ -T || --source-tarball ] ||		#
#				[ -v || --verbose ] ||			#
#				[ -V || --version ] ||			#
#				[ -- ]					#
#				[ non-option arguments ]		#
#									#
# Exit codes used:-							#
# Bash standard Exit Codes:	0 - success				#
#				1 - general failure			#
# User-defined exit code range is 64 - 113				#
#	C/C++ Semi-standard exit codes from sysexits.h range is 64 - 78	#
#		EX_USAGE	64	command line usage error	#
#		EX_DATAERR	65	data format error		#
#		EX_NOINPUT	66	cannot open input		#
#		EX_NOUSER	67	addressee unknown		#
#		EX_NOHOST	68	host name unknown		#
#		EX_UNAVAILABLE	69	service unavailable		#
#		EX_SOFTWARE	70	internal software error		#
#		EX_OSERR	71	system error (e.g., can't fork)	#
#		EX_OSFILE	72	critical OS file missing	#
#		EX_CANTCREAT	73	can't create (user) output file	#
#		EX_IOERR	74	input/output error		#
#		EX_TEMPFAIL	75	temp failure; user is invited	#
#					to retry			#
#		EX_PROTOCOL	76	remote error in protocol	#
#		EX_NOPERM	77	permission denied		#
#		EX_CONFIG	78	configuration error		#
#	User-defined (here) exit codes range 79 - 113:			#
#		None							#
#									#
# Further Info:								#
#									#
#########################################################################

#########################################################################
#									#
# Changelog								#
#									#
# Date		Author	Version	Description				#
#									#
# 25/06/2014	MG	1.0.1	Created.				#
# 02/08/2014	MG	1.0.2	Change naming from AutoTools to		#
#				AutoConf and Make.			#
# 27/10/2014	MG	1.0.3	Seperated each command in order to test	#
#				exit status.				#
# 13/11/2014	MG	1.0.4	Switched from getopts to GNU getopt to	#
#				allow long options.			#
# 16/11/2014	MG	1.0.5	Modify getopt processing to allow for	#
#				FreeBSD's quirk of 2 different getopt	#
#				programs. See comments at the start of	#
#				"Main"					#
# 16/11/2014	MG	1.0.6	Remove erroneous option.		#
# 18/11/2014	MG	1.0.7	Change FreeBSD specifics to *BSD and	#
#				change Linux to be the default.		#
# 24/11/2014	MG	1.0.8	Add overall package version to -V.	#
# 28/03/2015	MG	1.0.9	Remove BSD support.			#
# 31/01/2017	MG	1.1.1	Add debug make dist and make options.	#
# 25/02/2017	MG	1.2.0	Add distcheck & distcheckfake options.	#
# 25/06/2017	MG	1.2.1	Enforce 80 column rule.			#
# 01/12/2017	MG	1.2.2	Add SPDX license tags to source files.	#
# 04/12/2017	MG	1.2.3	Adopt standard exit codes; 0 on success	#
#				1 on failure.				#
# 06/02/2018	MG	1.3.1	Renamed from acmbuild.			#
#				Add -g option.				#
#				General script tidy up.			#
# 09/02/2018	MG	1.3.2	Remove script name from -V --version	#
#				print as this may have been invoked by	#
#				acmbuild.sh.				#
# 24/03/2018	MG	1.3.3	Add support for sparse CLA.		#
#				Add stderr log file.			#
# 07/04/2018	MG	1.3.4	Add -t --source-tarball CLA to build a	#
#				source tarball.				#
# 01/07/2018	MG	1.3.5	Separate configure from build actions	#
#				and make options more standardised.	#
# 06/08/2018	MG	1.3.6	Add -H --header-check option.		#
#				Change error log file to build log.	#
# 22/08/2018	MG	1.3.7	Add verbose option.			#
# 24/03/2019	MG	1.4.1	Style if, for and while loops after C.	#
#				Use -proper- booleans.			#
#				Remove script_exit_code variable 	#
#				propogating the exit code as a function	#
#				argument instead.			#
#				Improve trap exit code handling.	#
#				Re-factor into functions.		#
#				Introduce more meaningful exit codes.	#
#				Tidy up local and global variables.	#
#				Add -a --at-only option.		#
#				Add -t --testing-hacks option.		#
#				Add support for non-option arguments	#
#				which are not option arguments being	#
#				passed straight through to the		#
#				configure command line apart from the	#
#				first which is the base directory. This	#
#				allows support for things like		#
#				--prefix=... etc.			#
#				Add missing error check after getopt.	#
#				Replace #! env bash with absolute path	#
#				via configure.				#
# 05/04/2019	MG	1.4.2	Just execute getopt command AOT eval.	#
#				Setup trap as early as possible.	#
# 30/04/2019	MG	1.4.3	Correct getopt CL for proper quoting.	#
#				Ensure variables used as input to other	#
#				commands are inputised and evaluated	#
#				with eval.				#
# 18/06/2019	MG	1.4.4	Add -K --check option to run		#
#				make --quiet check.			#
# 25/06/2019	MG	1.4.5	Remove distcheckfake option. Now done	#
#				by distcheck with configure flags in	#
#				top level makefile.			#
# 28/10/2019	MG	1.4.6	Move script_exit() before it is used.	#
#				Cannot test for existence of file with	#
#				a variable which has retained quotes,	#
#				so introduce unquoted basedirunq.	#
# 01/12/2019	MG	1.4.7	Add parallel jobs option to pass to	#
#				make as --jobs				#
# 14/04/2021	MG	1.4.8	Add menu-config option to invoke menu	#
#				of configurable options.		#
# 28/05/2021	MG	1.4.9	Process menuconfig before using tee for	#
#				the build log as menuconfig may involve	#
#				dialog and redirection.			#
#				Use a temporary file to get result back	#
#				from configurable-options.sh		#
# 21/11/2021	MG	1.4.10	Tighten SPDX tag.			#
#									#
#########################################################################


##################
# Init variables #
##################

readonly version=1.4.10			# set version variable
readonly packageversion=1.3.12+rc1-6-g472cae210b7c	# Version of the complete package

# Set defaults
atonly=""
build=false
check=false
config=false
debug=""
dist=false
distcheck=false
gnulib=false
headercheck=""
menuconfig=false
par_jobs=""
sparse=""
tarball=false
testinghacks=""
verbose=false
verboseconfig=" --enable-silent-rules=yes"
verbosemake=" --quiet"
basedir="."			# Retain quotes
basedirunq=$basedir		# Without retaining quotes
configcli_extra_args=""


#############
# Functions #
#############

# -h --help output.
# No parameters
# No return value
usage ()
{
cat << EOF
Usage:- acmbuild.sh / $0 [options] [-- configure options to pass on]
	-a or --at-only during testing and for an AutoTools-only install, some
		build changes are required. e.g. You may reference an external
		Java jar in datadir but in AT builds and installations this
		may expand to /usr/local/share... So a substitution is required
		for this scenario.
	-b or --build make the project
	-c or --config congigure the project
	-C or --distcheck perform normal distcheck
	-d or --debug build with appropriate debug flags
	-D or --dist perform a make dist
	-g or --gnulib run gnulib-tool --update
	-h or --help displays usage information
	-H or --header-check show include stack depth
	-K or --check run make check
	-m or --menu-config Invoke menu of configurable options
	-p[X] or --parallel-jobs[=X] number of jobs to pass to make as --jobs=
		If not specified make is sequential
		If no value X is given then defaults to nproc
	-s or --sparse build using sparse
	-t or --testing-hacks some build changes may be required for testing
		purposes. e.g. A script may invoke a project jar file which
		when installed will be somewhere under datadir, but during
		testing it is in the project tree.
	-T or --source-tarball create source tarball
	-v or --verbose emit extra information
	-V or --version displays version information
EOF
}

# Standard function to emit messages depending on various parameters.
# Parameters -	$1 What:-	The message to emit.
#		$2 Where:-	stdout == 0
#				stderr == 1
# No return value.
output()
{
	if (( !$2 )); then
		printf "%s\n" "$1"
	else
		printf "%s\n" "$1" 1>&2
	fi
}

# Standard function to tidy up and return exit code.
# Parameters - 	$1 is the exit code.
# No return value.
script_exit()
{
	exit $1
}

# Standard function to test command error and exit if non-zero.
# Parameters - 	$1 is the exit code, (normally $? from the preceeding command).
# No return value.
std_cmd_err_handler()
{
	if (( $1 )); then
		script_exit $1
	fi
}

# Standard trap exit function.
# No parameters.
# No return value.
trap_exit()
{
	local -i exit_code=$?
	local msg

	msg="Script terminating with exit code $exit_code due to trap received."
	output "$msg" 1
	script_exit $exit_code
}

# Setup trap
trap trap_exit SIGHUP SIGINT SIGQUIT SIGTERM

# Process command line arguments with GNU getopt.
# Parameters -	$1 is the command line.
# No return value.
proc_CL()
{
	local GETOPTTEMP
	local script_name="acmbuild.sh/bootstrap.sh"
	local tmp

	tmp="getopt -o abcCdDghHKmp::stTvV "
	tmp+="--long at-only,build,check,config,distcheck,debug,dist,gnulib"
	tmp+=",help,header-check,menu-config,parallel-jobs::,sparse"
	tmp+=",source-tarball,testing-hacks,verbose,version"
	GETOPTTEMP=$($tmp -n "$script_name" -- "$@")
	std_cmd_err_handler $?

	eval set -- "$GETOPTTEMP"
	std_cmd_err_handler $?

	while true; do
		case "$1" in
		-a|--at-only)
			atonly=" --enable-atonly=yes"
			shift
			;;
		-b|--build)
			if $distcheck || $dist || $tarball; then
				msg="Options b, C, D and T are mutually "
				msg+="exclusive."
				output "$msg" 1
				script_exit 64
			fi
			build=true
			shift
			;;
		-c|--config)
			config=true
			shift
			;;
		-C|--distcheck)
			if $build || $dist || $tarball ; then
				msg="Options b, C, D and T are mutually "
				msg+="exclusive."
				output "$msg" 1
				script_exit 64
			fi
			distcheck=true
			shift
			;;
		-d|--debug)
			debug=" --enable-debug=yes"
			shift
			;;
		-D|--dist)
			if $build || $distcheck || $tarball ; then
				msg="Options b, C, D and T are mutually "
				msg+="exclusive."
				output "$msg" 1
				script_exit 64
			fi
			dist=true
			shift
			;;
		-g|--gnulib)
			gnulib=true
			shift
			;;
		-h|--help)
			usage
			shift
			script_exit 0
			;;
		-H|--header-check)
			headercheck=" --enable-headercheck=yes"
			shift
			;;
		-K|--check)
			if $distcheck || $dist || $tarball; then
				msg="Options C, D, K and T are mutually "
				msg+="exclusive."
				output "$msg" 1
				script_exit 64
			fi
			check=true
			shift
			;;
		-m|--menu-config)
			which whiptail > /dev/null
			status=$?
			if (( $status != 0 )); then
				output "Please first install whiptail." 1
				script_exit 64
			fi
			menuconfig=true
			shift
			;;
		-p|--parallel-jobs)
			if [[ -z "$2" ]]; then
				par_jobs=" --jobs=$(nproc)"
			else
				par_jobs=" --jobs=$2"
			fi
			shift 2
			;;
		-s|--sparse)
			sparse=" --enable-sparse=yes"
			shift
			;;
		-t|--testing-hacks)
			testinghacks=" --enable-testinghacks=yes"
			shift
			;;
		-T|--source-tarball)
			if $build || $distcheck || $dist ; then
				msg="Options b, C, D and T are mutually "
				msg+="exclusive."
				output "$msg" 1
				script_exit 64
			fi
			tarball=true
			shift
			;;
		-v|--verbose)
			verbose=true
			verboseconfig=""
			verbosemake=""
			shift
			;;
		-V|--version)
			printf "Script version %s\n" $version
			printf "Package version %s\n" $packageversion
			shift
			script_exit 0
			;;
		--)	shift
			break
			;;
		*)	output "Internal error." 1
			script_exit 64
			;;
		esac
	done

	if [[ $atonly || $debug || $headercheck || $sparse || $testinghacks ]] \
		|| $verbose ; then
		if ! $config ; then
			msg="Options a, d, H, s, t and v require option c."
			output "$msg" 1
			script_exit 64
		fi
	fi

	# One option has to be selected.
	if ! $build && ! $check && ! $config && ! $distcheck && ! $dist \
		&& ! $gnulib && ! $tarball ; then
		output "Either b, c, C, D, g, K or T must be set." 1
		script_exit 64
	fi

	# menu-config can only be selected with config
	if $menuconfig && ! $config; then
		output "Option c must be selected with option m." 1
		script_exit 64
	fi

	# First non-option argument which is not an option argument is the base
	# directory, all others are passed straight to the configure command
	# line, (to support things like  --prefix=... etc). Both of these need
	# to be inputised before they are passed on in order to maintain
	# original quoting. They can then be 'eval'ed.
	if (( $# )); then
		basedir=${1@Q}
		basedirunq="$1"		# Unquoted version
		shift
		configcli_extra_args+=" "${@@Q}
	fi
}

# Process gnulib option
# No parameters.
# No return value
proc_gnulib()
{
	local cmdline
	local msg
	local status

	if [[ -f "$basedirunq/m4/gnulib-cache.m4" ]]; then
		cmdline="gnulib-tool --update"$verbosemake$verbosemake
		cmdline+=" --dir="$basedir
		eval "$cmdline"
		status=$?
		output "$cmdline completed with exit status: $status" $status
		std_cmd_err_handler $status
	else
		msg="Option -g --gnulib ignored - "
		msg+="missing $basedir/m4/gnulib-cache.m4"
		output "$msg" 0
	fi
}

# Process the configurable options menu
# No parameters
# No return value
proc_menuconfig()
{
	local msg
	local readonly tmp_file=/tmp/$$.$(basename $0).tmp

	if [[ ! -f $basedirunq/configurable-options.sh \
		|| ! -r $basedirunq/configurable-options.sh \
		|| ! -x $basedirunq/configurable-options.sh ]]; then
		msg="The script configurable-options.sh must; exist in the"
		msg+=" project root directory, be readable and be executable."
		output "$msg" 1
		script_exit 77
	fi
	$basedirunq/configurable-options.sh $tmp_file
	std_cmd_err_handler $?
	configcli_extra_args+=$(cat < $tmp_file)
	std_cmd_err_handler $?
	rm -f $tmp_file
	std_cmd_err_handler $?
}

# Process configure
# No parameters
# No return value
proc_config()
{
	local cmdline
	local msg
	local status

	cmdline="autoreconf -if $basedir"
	eval "$cmdline"
	status=$?
	msg="autoreconf -if "$basedir" completed with exit status: $status"
	output "$msg" $status
	std_cmd_err_handler $status

	cmdline="$basedir/configure$configcli_extra_args$verboseconfig"
	cmdline+="$atonly$debug$headercheck$sparse$testinghacks"

	eval "$cmdline"
	status=$?
	output "$cmdline completed with exit status: $status" $status
	std_cmd_err_handler $status
}

# Process the correct make variation.
# No parameters.
# Returns 0 or command line error.
proc_make()
{
	local cmdline
	local status

	if $build ; then
		cmdline="make"$verbosemake$par_jobs
	fi

	if $check ; then
		if [[ ! $cmdline ]]; then
			cmdline="make"$verbosemake$par_jobs
		fi
		cmdline+=" check"
	fi

	if $distcheck ; then
		cmdline="make"$verbosemake$par_jobs" distcheck clean"
	fi

	if $dist ; then
		cmdline="make"$verbosemake$par_jobs" dist clean"
	fi

	if $tarball ; then
		cmdline="make"$verbosemake$par_jobs" srctarball clean"
	fi

	# May get here with cmdline empty if, for example, only the -g option
	# was set.
	if [[ ! $cmdline ]]; then
		return 0
	fi

	eval "$cmdline"
	status=$?
	output "$cmdline completed with exit status: $status" $status
	return $status
}


########
# Main #
########

proc_CL "$@"

# Now the main processing.
if $menuconfig; then
	proc_menuconfig
fi

# Create build log after menuconfig which may use dialog and redirections.
exec 1> >(tee build-output.txt) 2>&1

if $gnulib ; then
	proc_gnulib
fi

if $config ; then
	proc_config
fi

proc_make
std_cmd_err_handler $?

script_exit 0

